app/src/main/java/co/tinode/tindroid/MessagesAdapter.java

```diff
@@ -776,29 +878,34 @@ private void animateMessageBubble(final ViewHolder vh, boolean isMine, boolean l
 
     // Must match position-to-item of getItemId.
     private StoredMessage getMessage(int position) {
-        if (mCursor != null && !mCursor.isClosed()) {
-            return getMessage(mCursor, position, -1);
+        Cursor cursor = mCursor.getCursor();
+        if (cursor != null && !cursor.isClosed()) {
+            return position < mCursor.mMessages.size() ?
+                    mCursor.mMessages.get(position) :
+                    getMessage(cursor, position, -1);
         }
         return null;
     }
 
     @Override
     // Must match position-to-item of getMessage.
     public long getItemId(int position) {
-        if (mCursor != null && !mCursor.isClosed() && mCursor.moveToPosition(position)) {
-            return MessageDb.getLocalId(mCursor);
+        Cursor cursor = mCursor.getCursor();
+        if (cursor != null && !cursor.isClosed() && cursor.moveToPosition(position)) {
+            return MessageDb.getLocalId(cursor);
         }
         return View.NO_ID;
     }
 
     int findItemPositionById(long itemId, int first, int last) {
-        if (mCursor == null || mCursor.isClosed()) {
+        Cursor cursor = mCursor.getCursor();
```

Comment:
You sometimes check if mCursor is null and sometimes don't. Maybe double check?

Reply:
Done.

Reply:
I don't see it. Maybe you forgot to push the change?

Reply:
Nah, I misread your comment. Should be fixed now.